### PR TITLE
Support complex numbers in jax.scipy.signal.convolve/correlate

### PR DESCRIPTION
**Pull request by lukepfister**
Complex numbers are already supported in `jnp.convolve` and `jnp.correlate`.  This extends complex support to the n-dimensional `jax.scipy.signal` versions.

### CODE REVIEW

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

I'll wait for the CI to be green, then we can get this merged.

**Comment by lukepfister** [comment]
Huh.  The output is incorrectly conjugated in the numpy-dispatch build.  Not sure what's going on there, but I'll look into it.

**Comment by jakevdp** [comment]
That build also tests the oldest supported versions of numpy & scipy. Perhaps there was a bug in scipy convolve?

**Comment by jakevdp** [comment]
Wait, maybe the conjugate logic is just wrong for correlate - the other CIs were canceled because the numpy-dispatch one failed first.

**Comment by lukepfister** [comment]
Yeah, and they all pass on my end.  Very strange.  

```
[gw1] [ 10%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=convolve2d_xshape=int32[1,1]_yshape=int32[1,1]_mode=valid 
[gw0] [ 20%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=convolve2d_xshape=float32[2,2]_yshape=float32[2,3]_mode=same 
[gw2] [ 30%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=convolve2d_xshape=int32[2,2]_yshape=int32[1,1]_mode=full 
[gw9] [ 40%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=int32[3,4]_yshape=int32[3,4]_mode=same 
[gw6] [ 50%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=int32[2,2]_yshape=int32[2,3]_mode=valid 
[gw8] [ 60%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=int32[3,4]_yshape=int32[2,3]_mode=full 
[gw7] [ 70%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=int32[3,4]_yshape=int32[1,1]_mode=valid 
[gw5] [ 80%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=complex64[4,4]_yshape=complex64[2,3]_mode=valid 
[gw4] [ 90%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=complex64[4,4]_yshape=complex64[2,3]_mode=same 
[gw3] [100%] PASSED tests/scipy_signal_test.py::LaxBackedScipySignalTests::testConvolutions2Dop=correlate2d_xshape=complex64[3,4]_yshape=complex64[4,4]_mode=valid 

```

**Comment by jakevdp** [comment]
Can you try with the numpy/scipy versions used in that build? https://github.com/google/jax/blob/f6e0297c2224e84edaa52a2107878e4fc9578f38/.github/workflows/ci-build.yaml#L55

**Comment by lukepfister** [comment]
I can't get an install of Python 3.9, numpy 1.17.5, and scipy 1.2.1 to work-- conda doesn't have a numpy that old, and pip fails when building the scipy wheels.  I'll keep trying. 

On Python 3.8.10 with `numpy == 1.17.5` and `scipy == 1.2.1`, the tests pass locally.

**Comment by jakevdp** [comment]
The test in question is on Python 3.6 – I don't have any guesses as to why it's failing, but unfortunately we can't merge this PR with a broken test.

**Comment by jakevdp** [comment]
hmmm - I just thought of another thing: JAX's CI is run on the oldest supported jaxlib version, which is defined here: https://github.com/google/jax/blob/a82f0d8355715ee8015558a63c16d14a1659f102/jax/version.py#L16

It may be that complex convolutions are not correctly supported in that version of jaxlib.

**Comment by lukepfister** [comment]
> It may be that complex convolutions are not correctly supported in that version of jaxlib, and that you are using a newer jaxlib when you run the tests locally.

Afraid not-- I'm on jaxlib 0.1.65 locally.  

I think I'm getting somewhere, though.  The Python 3.7/x64=0 build passed, but the 3.9/x64=1 fails.  Running the tests locally with `JAX_ENABLE_X64=1` leads to the same style of test failures (incorrectly conjugated).

**Comment by jakevdp** [comment]
Ah, good. Another possibility: the logic may be always failing for certain input shape combinations, but those shape combinations only come up in certain test configurations. You could test that hypothesis by running with more generated cases; e.g.
```python
JAX_NUM_GENERATED_CASES=100 pytest -n auto tests/scipy_signal_test.py -k testConvolutions2D
```

**Comment by lukepfister** [comment]
I think I got the conjugation logic sorted out-- thanks for the pointer on increasing the number of tests.  Now we're failing with stuff like
```
E   Mismatched elements: 1 / 49 (2.04%)
E   Max absolute difference: 2.86428934e-14
E   Max relative difference: 1.38249275e-15
E    x: array([[ -0.668499  +1.332566j, -16.708536 -13.439743j,
E            22.589446  -4.512465j,  -2.600803 -13.421245j,
E            33.791187  +1.475926j, -20.786139  +4.966446j,...
E    y: array([[ -0.668499  +1.332566j, -16.708536 -13.439743j,
E            22.589446  -4.512465j,  -2.600803 -13.421245j,
E            33.791187  +1.475926j, -20.786139  +4.966446j,...
```
Very small errors that are just outside of the allowed tol/rtol limits.  I get these on my local machine, too.  Will look into it.

**Comment by jakevdp** [comment]
Great! Those sorts of errors are normal depending on the complexity of the computation involved. There are plenty of examples throughout the test files of changing the tolerances using a mapping of dtype to desired tolerance - let me know if you run into any questions surrounding that.

File: tests/scipy_signal_test.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def testConvolutions2D(self, xshape, yshape, dtype, mode, jsp_op, osp_op):
           jtu.format_shape_dtype_string(shape, dtype), axis, type, bp),
        "shape": shape, "dtype": dtype, "axis": axis, "type": type, "bp": bp}
       for shape in [(5,), (4, 5), (3, 4, 5)]
-      for dtype in default_dtypes
+      for dtype in jtu.dtypes.floating + jtu.dtypes.integer
```
Comment (jakevdp) [inline tests/scipy_signal_test.py:94]: Any reason for this change? `default_dtypes` tests `all_floating` and `all_integer`, so this change results in less test coverage.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great! One more comment on tests.

After that could you please squash your changes into a single commit? (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests) Then we'll be able to kick-off the merge.

Thanks for all the work on this!

**Comment by jakevdp** [comment]
I ran some additional jobs to confirm these changes pass tests on GPU & TPU. Nice work!

File: tests/scipy_signal_test.py (hunk 1 above)
Comment (lukepfister) [reply tests/scipy_signal_test.py:94]: Whoops, I missed this before the rebase/push, so let me get this fixed.

I'd like to exclude the complex types in `testDetrend`

Up at line 35 we have 
```python
default_dtypes = jtu.dtypes.floating + jtu.dtypes.integer + jtu.dtypes.complex
```
where I added the `jtu.dtypes.complex` part;  so I don't think that this module was testing `all_floating` or `all_integer` (at least how I understand it).

I could define a new `default_dtypes_plus_complex` and leave line 94 as `dtype in default_dtypes`; would that work?

**Comment by lukepfister** [comment]
> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogle%2Fjax%2Fpull%2F6954) for more info**.

Sorry about this, I accidentally pushed a commit from my work email (pfister@lanl.gov).  I removed the commit and pushed it again using my personal email.  I don't have a GitHub account associated with that email, but I do consent

File: tests/scipy_signal_test.py (hunk 1 above)
Comment (jakevdp) [reply tests/scipy_signal_test.py:94]: Oh I see - my mistake. I think this is fine as-is.

**Comment by jakevdp** [comment]
Looks like some other commits snuck in somehow. Can you pull down the `main` branch (recently renamed from `master`) and rebase on that?

**Review by jakevdp — APPROVED** [review]
Nice work!

**Comment by lukepfister** [comment]
> Nice work!

Thanks Jake. I really appreciate the guidance.
